### ISSUE TITLE
LBFGS optimization causes the value function model parameters to go to NAN

### ISSUE THREAD

**Issue by mjacar** [mjacar/pytorch-trpo#3]
Issue initially raised in #2. If you let the algorithm run sufficiently long with LBFGS value function optimization then you'll see EV_before and EV_after go to NAN. Investigating the root cause, it was found that this was caused by all the parameters of the value function model (self.value_function_model.state_dict()) going to NAN. Changing the optimizer to Adam in the ValueFunctionWrapper fixes this issue but makes the EV_before and EV_after numbers really crappy. The solution would be to look into how I'm using LBFGS wrong.

**Comment by AjayTalati** [comment]
Hi @mjacar, thanks a lot for your insight into this :+1: 

At the moment I haven't got any ideas on how to improve stability with LBFGS?

Thanks for looking into using Adam, I'll try RMSprop and see if it helps?

**Comment by AjayTalati** [comment]
Hi @mjacar, seems RMSprop works quite well :+1: 

```
Evaluation avg reward = 466.880000 
KL_Old_New: 0.0068593756295740604
Entropy: 0.5847200751304626
EV_Before: 0.36330654205668456
EV_After: 0.3662000964838015
Evaluation avg reward = 471.130000 
KL_Old_New: 0.005557108670473099
Entropy: 0.5804362893104553
EV_Before: 0.5606148255500768
EV_After: 0.7753341472443888
Evaluation avg reward = 432.330000 
KL_Old_New: 0.005399847403168678
Entropy: 0.5831596255302429
EV_Before: 0.7288128439943649
EV_After: 0.6781972111188941
Evaluation avg reward = 458.860000 
KL_Old_New: 0.007636414375156164
Entropy: 0.5718984603881836
EV_Before: 0.44457040021299243
EV_After: 0.8350147280139323
Evaluation avg reward = 485.190000 
KL_Old_New: 0.00448857294395566
Entropy: 0.5903996229171753
EV_Before: 0.585823396053786
EV_After: 0.3368593059093885
Evaluation avg reward = 435.050000 
KL_Old_New: 0.007822290994226933
Entropy: 0.5978102684020996
EV_Before: 0.40117586783231374
EV_After: 0.5880426395447291
```

**Comment by mjacar** [comment]
LBFGS divergence can be resolved by just making the learning rate smaller.

**Comment by AjayTalati** [comment]
Thanks - what learning rate are you now using to avoid the divergence?

**Comment by mjacar** [comment]
In the new example I just pushed, the default learning_rate of 1 seems to work from what I've seen (though I haven't run the example for a long time yet). In the old example, I think I got things working fine with a learning rate of around 0.25.
